src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        for (let j = 0; j < action.keysToRemove.length; j++) {
+          const key = action.keysToRemove[j];
+
+          delete currentState[key];
+        }
+        break;
+
+      case 'clear':
+        currentState = {};
+        break;
+    }
+    stateHistory.push(currentState);
```

Comment:
You are pushing the reference to 'currentState' into 'stateHistory' after each action. This will result in all elements of 'stateHistory' pointing to the same object, which is the final state. To fix this, you should push a copy of 'currentState' instead to ensure each element in 'stateHistory' is a unique object representing the state at that point in time.